File: src/main/java/com/example/TableDataFetcher.java
package com.example;

import com.example.calcitewrappers.DatabaseManager;
import com.example.calcitewrappers.ForeignKey;
import com.example.calcitewrappers.ForeignKeyManager;
import com.example.calcitewrappers.FullyQualifiedTableName;
import graphql.schema.*;
import org.apache.calcite.rel.RelNode;
import org.apache.calcite.rex.RexLiteral;
import org.apache.calcite.rex.RexNode;
import org.apache.calcite.tools.RelBuilder;
import org.apache.calcite.tools.RelRunner;
import org.apache.calcite.util.Pair;
import org.jetbrains.annotations.Nullable;

import java.sql.PreparedStatement;
import java.sql.SQLException;
import java.util.*;
import java.util.stream.Stream;

import static com.example.Utils.reverse;
import static java.util.Objects.requireNonNull;

public final class TableDataFetcher implements DataFetcher {

    private final FullyQualifiedTableName fullyQualifiedTableName;
    private final SqlToGraphQLConfiguration sqlToGraphQLConfiguration;
    private final DatabaseManager databaseManager;

    private final String databaseName;
    private final String schemaName;
    private final String tableName;

    private final Set<ForeignKey> foreignKeys;

    public TableDataFetcher(
            FullyQualifiedTableName fullyQualifiedTableName,
            SqlToGraphQLConfiguration sqlToGraphQLConfiguration,
            DatabaseManager databaseManager) {
        this.fullyQualifiedTableName = fullyQualifiedTableName;
        this.sqlToGraphQLConfiguration = sqlToGraphQLConfiguration;
        this.databaseManager = databaseManager;

        this.databaseName = fullyQualifiedTableName.database();
        this.schemaName = fullyQualifiedTableName.schema();
        this.tableName = fullyQualifiedTableName.table();

        this.foreignKeys = ForeignKeyManager.getForeignKeysForTable(fullyQualifiedTableName);
    }


    @Override
    public Object get(DataFetchingEnvironment environment) throws Exception {
        return handleField(environment.getArguments(), environment.getSelectionSet(), null);
    }

    private List<Map<String, Object>> handleField(
            Map<String, Object> arguments,
            DataFetchingFieldSelectionSet selectionSet,
            @Nullable RexNode joinFilterCondition
    ) throws SQLException {
        Map<String, Object> where = (Map<String, Object>) arguments.get("where");
        Integer limit = (Integer) arguments.get("limit");
        Integer offset = (Integer) arguments.get("offset");

        if (limit == null) {
            limit = 0;
        }
        if (offset == null) {
            offset = 0;
        }

        List<SelectedField> scalars = new ArrayList<>();
        List<SelectedField> joins = new ArrayList<>();

        for (SelectedField field : selectionSet.getImmediateFields()) {
            if (field.getType() instanceof GraphQLScalarType) {
                scalars.add(field);
            } else {
                joins.add(field);
            }
        }

        RelNode relNode = buildRelationalExpr(joinFilterCondition, where, offset, limit, scalars);
        List<Map<String, Object>> rows;
        RelRunner relRunner = this.databaseManager.getRelRunner();
        try (PreparedStatement statement = relRunner.prepareStatement(relNode)) {
            rows = SQLUtils.executeStatementReturningListOfObjects(statement);
        }

        for (SelectedField join : joins) {
            var foreignKey = foreignKeys.stream()
                    .filter(fk -> fk.sourceTable().table().equals(join.getName()) || fk.targetTable().table().equals(join.getName()))
                    .findFirst()
                    .orElseThrow(() -> new RuntimeException("Could not find foreign key for " + join.getName()));

            boolean tableIsTargetInJoin = foreignKey.targetTable() == fullyQualifiedTableName;
            FullyQualifiedTableName joinTable = tableIsTargetInJoin ? foreignKey.sourceTable() : foreignKey.targetTable();

            RelBuilder innerRelBuilder = databaseManager.getRelBuilder();
            innerRelBuilder = innerRelBuilder.scan(joinTable.database(), joinTable.schema(), joinTable.table());

            List<RexNode> joinConditions = new ArrayList<>();
            for (Pair<String, String> colPair : foreignKey.columns()) {
                List<RexLiteral> list = new ArrayList<>();
                if (tableIsTargetInJoin) {
                    for (Map<String, Object> row : rows) {
                        RexLiteral literal = innerRelBuilder.literal(row.get(colPair.right));
                        list.add(literal);
                    }
                    joinConditions.add(innerRelBuilder.in(innerRelBuilder.field(colPair.left), list));
                } else {
                    for (Map<String, Object> row : rows) {
                        RexLiteral literal = innerRelBuilder.literal(row.get(colPair.left));
                        list.add(literal);
                    }
                    joinConditions.add(innerRelBuilder.in(innerRelBuilder.field(colPair.right), list));
                }
            }

            List<Map<String, Object>> joinResults = handleField(
                    arguments,
                    join.getSelectionSet(),
                    innerRelBuilder.and(joinConditions)
            );

            if (tableIsTargetInJoin) {
                for (Map<String, Object> row : rows) {
                    row.put(join.getName(),
                            joinResults.stream()
                                    .filter(joinRow ->
                                            foreignKey.columns().stream().allMatch(colPair -> {
                                                return joinRow.get(colPair.left).equals(row.get(colPair.right));
                                            })
                                    )
                    );
                }
            } else {
                for (Map<String, Object> row : rows) {
                    row.put(join.getName(),
                            joinResults.stream()
                                    .filter(joinRow ->
                                            foreignKey.columns().stream().allMatch(colPair -> {
                                                return joinRow.get(colPair.right).equals(row.get(colPair.left));
                                            })
                                    )
                                    .findFirst()
                                    .orElse(null)
                    );
                }
            }
        }

        return rows;
    }

    private RelNode buildRelationalExpr(
            @Nullable RexNode joinFilterCondition,
            Map<String, Object> where,
            Integer offset,
            Integer limit,
            List<SelectedField> scalars
    ) {
        RelBuilder relBuilder = databaseManager.getRelBuilder();

        if (joinFilterCondition == null) {
            if (schemaName != null) {
                relBuilder = relBuilder.scan(databaseName, schemaName, tableName);
            } else {
                relBuilder = relBuilder.scan(databaseName, tableName);
            }
        }

        if (where != null) {
            Expression whereExpr = whereArgumentToExpression(where, scalars);
            if (whereExpr != null) {
                relBuilder = relBuilder.filter(whereExpr.toRexNode(relBuilder));
            }
        }

        if (joinFilterCondition != null) {
            relBuilder = relBuilder.filter(joinFilterCondition);
        }

        if (offset != 0 || limit != 0) {
            relBuilder = relBuilder.limit(offset, limit == 0 ? -1 : limit);
        }

        List<String> columns = scalars.stream().map(SelectedField::getName).toList();
        relBuilder = relBuilder.project(relBuilder.fields(columns));

        return relBuilder.build();
    }

    public Expression whereArgumentToExpression(Map<?, ?> whereArgument, List<SelectedField> fields) {
        // reduceRight
        Stream<Expression> expr = whereArgument.entrySet().stream().map((entry) -> {
            String key = (String) requireNonNull(entry.getKey());
            Object value = requireNonNull(entry.getValue());
            return switch (key) {
                case "_and" -> {
                    if (value instanceof List<?> list) {
                        yield reverse(list)
                                .map(e -> whereArgumentToExpression((Map) e, fields))
                                .reduce(Expression.AND::new)
                                .orElse(null);
                    }
                    throw new IllegalArgumentException("Expected list of maps");
                }
                case "_or" -> {
                    if (value instanceof List<?> list) {
                        yield reverse(list)
                                .map(e -> whereArgumentToExpression((Map) e, fields))
                                .reduce(Expression.OR::new)
                                .orElse(null);
                    }
                    throw new IllegalArgumentException("Expected list of maps");
                }
                case "_not" -> {
                    if (value instanceof Map m) {
                        yield new Expression.NOT(whereArgumentToExpression(m, fields));
                    }
                    throw new IllegalArgumentException("Expected map");
                }
                default -> {
                    if (value instanceof Map m) {
                        assert m.size() == 1 : "Expected single operator";

                        GraphQLInputTypeForScalar graphQLInputTypeForScalar = sqlToGraphQLConfiguration
                                .getGraphqlScalarTypeToGraphQLInputTypeMap()
                                .get(getScalarTypeOfField(key, fields));

                        if (graphQLInputTypeForScalar != null) {
                            String operator = (String) m.keySet().iterator().next();
                            Object operand = m.values().iterator().next();
                            QueryFilterExpression queryFilterExpression = new QueryFilterExpression(operator, key, operand);
                            yield graphQLInputTypeForScalar.handleQueryFilterExpression(queryFilterExpression);
                        }

                        throw new IllegalArgumentException("Expected scalar type");
                    }
                    throw new IllegalArgumentException("Expected map");
                }
            };
        });

        return reverse(expr.toList()).reduce(Expression.AND::new).orElse(null);
    }


    @Nullable GraphQLScalarType getScalarTypeOfField(String name, List<SelectedField> fields) {
        SelectedField field = fields.stream().filter(f -> f.getName().equals(name)).findFirst().orElse(null);
        if (field == null) {
            return null;
        }

        GraphQLType fieldType = field.getType();
        if (fieldType instanceof GraphQLScalarType scalarType) {
            return scalarType;
        }

        return null;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == this) return true;
        if (obj == null || obj.getClass() != this.getClass()) return false;
        var that = (TableDataFetcher) obj;
        return Objects.equals(this.fullyQualifiedTableName, that.fullyQualifiedTableName) &&
                Objects.equals(this.sqlToGraphQLConfiguration, that.sqlToGraphQLConfiguration) &&
                Objects.equals(this.databaseManager, that.databaseManager);
    }

    @Override
    public int hashCode() {
        return Objects.hash(fullyQualifiedTableName, sqlToGraphQLConfiguration, databaseManager);
    }

    @Override
    public String toString() {
        return "TableDataFetcher[" +
                "fullyQualifiedTableName=" + fullyQualifiedTableName + ", " +
                "sqlToGraphQLConfiguration=" + sqlToGraphQLConfiguration + ", " +
                "calciteSchemaManager=" + databaseManager + ']';
    }


}

File: src/main/kotlin/TableDataFetcher.kt
import extensions.toListOfMaps
import graphql.schema.DataFetcher
import graphql.schema.DataFetchingEnvironment
import graphql.schema.DataFetchingFieldSelectionSet
import graphql.schema.GraphQLScalarType
import graphql.schema.SelectedField
import org.apache.calcite.rel.RelNode
import org.apache.calcite.rex.RexNode

data class TableDataFetcher(val fullyQualifiedTableName: FullyQualifiedTableName) :
    DataFetcher<Any> {
    private val databaseName = fullyQualifiedTableName.database
    private val schemaName = fullyQualifiedTableName.schema
    private val tableName = fullyQualifiedTableName.table
    private val foreignKeys = ForeignKeyManager.getForeignKeysForTable(fullyQualifiedTableName)

    override fun get(env: DataFetchingEnvironment): Any {
        return mutuallyRecursiveHandleField(env.arguments, env.selectionSet)
    }

    /**
     * For each query field:
     *  1. Create a query by:
     *      a. Scan on (DB, Schema, Table)
     *      b. Project all scalar fields
     *      c. Convert the "where" argument to a filter clause
     *      d. Add "limit" and "offset" clauses
     *  2. Run the query
     *  3. For each nested field selection in the query:
     *      a. Repeat the above (Adding the foreign key to the filter clause
     *         as an IN statement using the result of the previous query)
     *      b. Add the results to the parent record under the nested field's name
     */
    fun mutuallyRecursiveHandleField(
        arguments: Map<String, Any>,
        selectionSet: DataFetchingFieldSelectionSet,
        joinFilterCondition: RexNode? = null
    ): List<Map<String, Any>> {
        val where = arguments["where"] as Map<String, Any>?
        val limit = arguments["limit"] as Int?
        val offset = arguments["offset"] as Int?
        val orderBy = arguments["order_by"] as List<String>?

        val (scalars, joins) = selectionSet.immediateFields.partition {
            it.type is GraphQLScalarType
        }

        val relNode = buildRelationalExpr(joinFilterCondition, where, offset, limit, scalars)
        val rows = CalciteSchemaManager.executeQuery(relNode).toListOfMaps()

        for (join in joins) {
            // Recursively mutates the row values in parent records to attach values from child tables
            // TODO: Can this be immutable?
            mutuallyRecursiveHandleJoinField(join, rows)
        }

        return rows
    }

    private fun buildRelationalExpr(
        joinFilterCondition: RexNode?,
        where: Map<String, Any>?,
        offset: Int?,
        limit: Int?,
        scalars: List<SelectedField>
    ): RelNode {
        val relBuilder = CalciteSchemaManager.relBuilder.apply {
            // If there is a filter condition, we already have a scan on the stack and so shouldn't create a new one
            if (joinFilterCondition == null) {
                // Scan on (DB, Schema, Table)
                if (schemaName != null)
                    this.scan(databaseName, schemaName, tableName)
                else
                    this.scan(databaseName, tableName)
            }

            //  Convert the "where" argument to a filter clause
            if (where != null)
                this.filter(whereArgumentToExpression(where).toRexNode(this))

            // If there is a join filter condition, we need to add it to the filter clause
            if (joinFilterCondition != null)
                this.filter(joinFilterCondition)

            // Add "limit" and "offset" clauses
            if (offset != null || limit != null)
                this.limit((offset ?: 0), (limit ?: -1))

            // Project all scalar fields
            val columns = scalars.map { this.field(it.name) }
            this.project(columns)
        }

        return relBuilder.build()
    }

    // Mutually recursive with "mutuallyRecursiveHandleField"
    private fun mutuallyRecursiveHandleJoinField(
        join: SelectedField,
        rows: MutableList<MutableMap<String, Any>>
    ) {
        val foreignKey = foreignKeys.find { it.sourceTable.table == join.name || it.targetTable.table == join.name }
        if (foreignKey == null) return

        val tableIsTargetInJoin = foreignKey.targetTable == fullyQualifiedTableName
        val joinTable = if (tableIsTargetInJoin) foreignKey.sourceTable else foreignKey.targetTable

        val innerRelBuilder = CalciteSchemaManager.relBuilder
        innerRelBuilder.scan(joinTable.database, joinTable.schema, joinTable.table)

        val joinResults = mutuallyRecursiveHandleField(
            join.arguments,
            join.selectionSet,
            innerRelBuilder.and(
                foreignKey.columns.map { colPair ->
                    if (tableIsTargetInJoin) {
                        innerRelBuilder.`in`(
                            innerRelBuilder.field(colPair.first),
                            rows.map { innerRelBuilder.literal(it[colPair.second]) }
                        )
                    } else {
                        innerRelBuilder.`in`(
                            innerRelBuilder.field(colPair.second),
                            rows.map { innerRelBuilder.literal(it[colPair.first]) }
                        )
                    }
                }
            )
        )

        // TODO: Check if the below is faster:
        //  - Grouping the rows by their primary key first
        //  - Then looking up directly instead of filtering
        //  - Finally flattening
        fun buildParentRowProperty(foreignKey: ForeignKey) {
            if (tableIsTargetInJoin) {
                for (row in rows) {
                    row[join.name] = joinResults.filter { joinRow ->
                        foreignKey.columns.all { colPair ->
                            joinRow[colPair.first] == row[colPair.second]
                        }
                    }
                }
            } else {
                for (row in rows) {
                    row[join.name] = joinResults.first { joinRow ->
                        foreignKey.columns.all { colPair ->
                            joinRow[colPair.second] == row[colPair.first]
                        }
                    }
                }
            }
        }

        buildParentRowProperty(foreignKey)
    }

    private fun whereArgumentToExpression(whereArgument: Map<*, *>): Expression {
        val exprs = whereArgument.entries.map { (key, value) ->
            when (key) {
                "_and" -> {
                    if (value is List<*>) {
                        value.map {
                            if (it is Map<*, *>) {
                                return whereArgumentToExpression(it)
                            } else {
                                throw IllegalArgumentException("_and must be a list of expressions")
                            }
                        }.reduceRight(::AND)
                    }
                    throw IllegalArgumentException("_and must be a list")
                }
                "_or" -> {
                    if (value is List<*>) {
                        value.map {
                            if (it is Map<*, *>) {
                                return whereArgumentToExpression(it)
                            } else {
                                throw IllegalArgumentException("_and must be a list of expressions")
                            }
                        }.reduceRight(::OR)
                    }
                    throw IllegalArgumentException("_or must be a list")
                }
                "_not" -> {
                    if (value is Map<*, *>) {
                        return NOT(whereArgumentToExpression(value))
                    } else {
                        throw IllegalArgumentException("_not must be an expression")
                    }
                }
                else -> {
                    if (value is Map<*, *>) {
                        assert(value.size == 1) { "Expected single value" }
                        if (key is String) {
                            return getExpression(key, value)
                        }
                        throw IllegalArgumentException("Expected string key")
                    }
                    throw IllegalArgumentException("Unknown key $key")
                }
            }
        }
        return exprs.reduceRight(::AND)
    }

    private fun getExpression(column: String, valueMap: Map<*, *>): Expression {
        val operator = valueMap.keys.iterator().next()
        val operand = requireNotNull(valueMap.values.iterator().next()) { "Operand must be present" }
        return when (operator) {
            "_eq" -> EQ(COLUMN(column), LITERAL(operand))
            "_neq" -> NEQ(COLUMN(column), LITERAL(operand))
            "_gt" -> GT(COLUMN(column), LITERAL(operand))
            "_gte" -> GTE(COLUMN(column), LITERAL(operand))
            "_lt" -> LT(COLUMN(column), LITERAL(operand))
            "_lte" -> LTE(COLUMN(column), LITERAL(operand))
            "_in" -> IN(COLUMN(column), LITERAL(operand))
            "_nin" -> NIN(COLUMN(column), LITERAL(operand))
            "_is_null" -> IS_NULL(COLUMN(column))
            else -> throw IllegalArgumentException("Unknown operator: $operator")
        }
    }
}
